cabal-install/src/Distribution/Client/CmdStatus.hs

```diff
@@ -0,0 +1,367 @@
+
+data StatusOutputFormat
+  = JSON
+  deriving (Eq, Ord, Show, Read)
+
+data StatusFlags = StatusFlags
+  { statusBuildInfo :: [String]
+  , statusCompiler :: Flag Bool
+  , statusOutputFormat :: Flag StatusOutputFormat
+  } deriving (Eq, Show, Read)
+
+defaultStatusFlags :: StatusFlags
+defaultStatusFlags = StatusFlags
+  { statusBuildInfo = mempty
+  , statusCompiler = mempty
+  , statusOutputFormat = mempty
+  }
+
+statusOutputFormatParser :: ReadE (Flag StatusOutputFormat)
+statusOutputFormatParser = ReadE $ \case
+  "json" -> Right $ Flag JSON
+  policy -> Left  $ "Cannot parse the status output format '"
+            <> policy <> "'"
+
+statusOutputFormatPrinter
+  :: Flag StatusOutputFormat -> [String]
+statusOutputFormatPrinter = \case
+  (Flag JSON) -> ["json"]
+  NoFlag      -> []
+
+statusOptions :: ShowOrParseArgs -> [OptionField StatusFlags]
+statusOptions showOrParseArgs =
+  [ option [] ["output-format"]
+    "Output Format for the information"
+    statusOutputFormat (\v flags -> flags { statusOutputFormat = v })
+    (reqArg "json"
+      statusOutputFormatParser
+      statusOutputFormatPrinter
+    )
+  , option [] ["build-info"]
+    "List all available targets in the project"
+    statusBuildInfo (\v flags -> flags { statusBuildInfo = v ++ statusBuildInfo flags})
+    (reqArg "TARGET" buildInfoTargetReadE (fmap show))
+  , option [] ["compiler-info"]
+    "Print information of the project compiler"
+    statusCompiler (\v flags -> flags { statusCompiler = v })
+    (yesNoOpt showOrParseArgs)
+  ]
+  where
+    buildInfoTargetReadE :: ReadE [String]
+    buildInfoTargetReadE =
+      parsecToReadE
+        -- This error should never be shown
+        ("couldn't parse targets: " ++)
+        -- TODO: wrong parser, kills filepaths with spaces
+        (parsecCommaList parsecToken)
+
+-------------------------------------------------------------------------------
+-- Action
+-------------------------------------------------------------------------------
+
+-- | Entry point for the 'ide' command.
+statusAction :: NixStyleFlags StatusFlags -> [String] -> GlobalFlags -> IO ()
+statusAction flags@NixStyleFlags { extraFlags = statusFlags, ..} cliTargetStrings globalFlags = do
+  when (NoFlag == statusOutputFormat statusFlags) $ do
+    die' verbosity "The status command requires the flag '--output-format'."
+  when (not $ null cliTargetStrings) $
+    die' verbosity "The status command takes not target arguments directly. Use appropriate flags to pass in target information."
+
+  baseCtx <- establishProjectBaseContext verbosity cliConfig OtherCommand
+  (_, elaboratedPlan, elabSharedConfig, _, _) <-
+    rebuildInstallPlan verbosity
+                        (distDirLayout baseCtx)
+                        (cabalDirLayout baseCtx)
+                        (projectConfig baseCtx)
+                        (localPackages baseCtx)
+
+  let initialJson = Json.object
```

Comment:
Shouldn't we already structure this function in a way that the information is decoupled from the formatting?

Return a big product type and then allow to feed that into the formatter function.

Reply:
Changed, better now?